docs(headless,headless-react): typedoc styling update and nav cookies as functional

jsmith2-coveo opened the pull request:
https://coveord.atlassian.net/browse/KIT-3964

Based on compliance for cookies, an update was made to TypeDoc to allow for disabling of local storage usage, and is contained in v0.28.1. This PR addresses the update to v0.28.1, and incorporates the change to disable local storage usage by typedoc features when functional cookies are disabled, respecting compliance requirements.

It also updates the styling of typedoc generated sites to better follow the style present on the main docs site. The header bar now contains the same functionality as our main docs site and the style of the body is more consistent with our main docs also.

alexprudhomme commented:
Why is there 60k lines of css in `main-new.css` ? How did you create that file ?

jsmith2-coveo commented:
> Why is there 60k lines of css in `main-new.css` ? How did you create that file ?

It's a copy of the CSS files from the `docs.coveo.com`. I made the decision to copy it here rather than referencing it via a link, because if we make changes to the CSS specifically for the `docs.coveo.com` site we could have breaking effects on the generated TypeDoc page. 

The goal was to have the TypeDoc site closer to the style of the `docs.coveo.com` site, and the consensus was that cherry picking out the CSS specific to the TypeDoc generated page from the `main-new.css` file would take more time than it would be worth. 

Happy to discuss further, or adjust accordingly though!

jpmarceau commented on packages/headless/typedoc/assets/vars/dark-mode-toggle.js:
```diff
@@ -0,0 +1,360 @@
+    });
+  }
+
+  attributeChangedCallback(name, oldValue, newValue) {
```
You don't seem to use `oldValue`
```suggestion
  attributeChangedCallback(name, newValue) {
```

jsmith2-coveo replied:
Good catch! Thanks JP!

louis-bompart replied:
⚠️ the suggested code is invalid. 
```suggestion
  attributeChangedCallback(name, _oldValue, newValue) {
```

You can't just "omit" a parameter; order matters.

jpmarceau commented on packages/headless-react/typedoc/assets/vars/dark-mode-toggle.js:
```diff
@@ -0,0 +1,360 @@
+    });
+  }
+
+  attributeChangedCallback(name, oldValue, newValue) {
```
You don't seem to use `oldValue`
```suggestion
  attributeChangedCallback(name, newValue) {
```

jsmith2-coveo replied:
Good catch! Thanks JP!

louis-bompart replied:
⚠️ the suggested code is invalid. 
```suggestion
  attributeChangedCallback(name, _oldValue, newValue) {
```

You can't just "omit" a parameter; order matters.

jpmarceau reviewed (approved):
Great work Josh!
I learned a number of things just reviewing this PR.

spiracha-coveo reviewed (commented):
Since some of the functionality is being reused between headless and headless-react, is there a way to share it instead of duplicating it in both places? Future maintenance would be easier if we don’t violate the DRY principle

jpmarceau commented:
> Since some of the functionality is being reused between headless and headless-react, is there a way to share it instead of duplicating it in both places? Future maintenance would be easier if we don’t violate the DRY principle

I remember looking into this with @alexprudhomme initially when setting up both headless and headless-react typedoc projects. It wasn't possible to merge the projects and we decided to duplicate code, although less code was involved at the time... But at the number of files we reach here, I see your point. Maybe we could centralize the common files, and copy them dynamically in the headless/headless-react project when executing the `build:typedoc` command?

jsmith2-coveo commented:
I've moved a large number of the files to a `shared-typedoc` directory structure and updated the script for how they get moved to simplify the process.
